### PR TITLE
341 tabnav child info

### LINKED ISSUE #341: Missing info icon on child info tab sections
Screen Shot 2020-09-18 at 8 05 39 AM
See also #287 for other tab-nav styling clean-ups

### PR DESCRIPTION
**Pull request by mmazanec22**
Closes #341

### CODE REVIEW

File: client/src/utils/getValidationStatus.ts [hunk 1]
```diff
@@ -57,13 +57,21 @@ export function getValidationStatusForField<
...
   fields: string[],
   options?: ValidationStatusOptions
 ): FormStatusProps | undefined {
+  if (Array.isArray(data)) {
+    const allErrors = data.map((d: T) =>
+      getValidationStatusForFields(d, fields, options)
+    );
+    const anError = allErrors.find((e) => !!e);
+    // Assumes that the error for one of the objects in the array is the same for all
+    return anError || undefined;
```
Comment (bamader) [inline client/src/utils/getValidationStatus.ts:73]: This assumption should always hold, right?

File: src/utils/distributeValidationErrorsToSubObjects.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  copiedParent.validationErrors
+    .filter((v: ValidationError) => v.children?.length)
+    .forEach((parentValidationError: ValidationError) => {
+      let childObject = copiedParent[parentValidationError.property];
+      if (Array.isArray(childObject)) {
+        // i.e. income determinations or enrollments
+        childObject = childObject.map((childObjectItem, i) => {
+          const childError = parentValidationError.children.find(e => e.property === `${i}`)
+          childObjectItem.validationErrors = childError.children;
+          return distributeValidationErrorsToSubObjects(childObjectItem);
```
Comment (bamader) [inline src/utils/distributeValidationErrorsToSubObjects.ts:25]: Gotta love that sweet sweet recursion 😛

**Review by bamader — APPROVED** [review]
LGTM! Only question is on the default assumption that if a child object doesn't exist we say that a form field component has errors (like in this example `child ? !!getValidationStatusForFields(child, childIdentifiersFields) : true`). I'm assuming this is because if a child or one of its nested things doesn't exist, then we want to flag errors for that component anyway?

File: client/src/containers/AddChild/AddChild.tsx [hunk 3]
```diff
@@ -46,46 +46,33 @@ const AddChild: React.FC = () => {
...
-  }, [
-    accessToken,
-    child,
-    locationState,
-    organization,
-    history,
-    updateChild,
-    creating,
-  ]);
+      .finally(triggerRefetchChild);
```
Comment (jlhogan) [inline client/src/containers/AddChild/AddChild.tsx:72]: why does the child need to be refetched on failed post?

File: client/src/routes.ts [hunk 4]
```diff
@@ -69,6 +69,11 @@ export const routes: RouteConfig[] = [
     unauthorized: false,
     exact: true,
   },
+  {
+    path: '/create-record/:childId',
+    component: AddChild,
+    unauthorized: false,
+  },
   {
     path: '/create-record',
```
Comment (jlhogan) [inline client/src/routes.ts:78]: i _think_ you can do something like `path: '/create-record/:childId?` to indicate the thing is optional and it'll match both

File: client/src/utils/getValidationStatus.ts (hunk 1 above)
Comment (jlhogan) [reply client/src/utils/getValidationStatus.ts:73]: when is the data an array ?

File: src/controllers/children.ts [hunk 5]
```diff
@@ -67,7 +68,7 @@ export const getChildById = async (id: string) => {
       return propertyDateSorter(enrollmentA, enrollmentB, (e) => e.exit);
     });
   }
-  return { ...child, validationErrors: await validate(child) };
+  return distributeValidationErrorsToSubObjects({ ...child, validationErrors: await validate(child, { validationError: { target: false } }) });
```
Comment (jlhogan) [inline src/controllers/children.ts:71]: nit can we change the function signature of this to take in the child and the validation errors, as opposed to them already composed? just for like nice to look at ness?

**Review by jlhogan — APPROVED** [review]
barring the open Q about validation errors on array data

File: client/src/components/Forms/FamilyAddress/Form.tsx [hunk 6]
```diff
@@ -57,7 +62,7 @@ export const FamilyAddressForm: React.FC<EditFormProps> = ({
     <div className="grid-container margin-top-2">
       {!hideHeader && <h2 className="grid-row">Family Address</h2>}
       <Form<Family>
-        className="FamilyInfoForm"
+        className="FamilyAddressForm"
```
[the remark this answers is not in the document]
Comment (mmazanec22) [reply client/src/components/Forms/FamilyAddress/Form.tsx:65]: Nope, besides make snapshots easier to look at

File: client/src/utils/getValidationStatus.ts (hunk 1 above)
Comment (mmazanec22) [reply client/src/utils/getValidationStatus.ts:73]: Family income dets, child enrollments

**Comment by mmazanec22** [comment]
@jlhogan I merged this before seeing your comments but I'll make a fast follow PR to address them

**Comment by jlhogan** [comment]
word! I tend to approve unless something is NO WAY!!!!!!!!!!!! but if it's more helpful to just add comments if there are still open questions, I can definitely do that instead

**Pull request merged**